### ISSUE TITLE
Merge commits break LLVM CI download

### ISSUE THREAD

**Issue by jachris** [rust-lang/rust#101907]
While PRs are not allowed to have merge commits, they can still be useful locally (or for not rewriting history for longer-running branches). 

However, they break the commit detection logic for downloading CI artifacts.

### Steps to Reproduce

1. `git checkout <old-commit>`
2. `git merge --no-ff master`
3. `./x.py build library/std`

### Expected Result

Builds fine.

### Actual Result

~~~
...
downloading https://ci-artifacts.rust-lang.org/rustc-builds/<old-commit>/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404      #                                                                               

error: failed to download llvm from ci

help: old builds get deleted after a certain time
help: if trying to compile an old commit of rustc, disable `download-ci-llvm` in config.toml:

[llvm]
download-ci-llvm = false

Build completed unsuccessfully in 0:01:43
~~~

**Comment by Mark-Simulacrum** [comment]
Can you say more about their usefulness? What are you merging in? We can maybe support merging non-bors commits, though it brings some complexity (e.g., which parent do we follow for the "nearest" bors commit, and what does nearest mean at that point), so in general I would be inclined to say this isn't supported and you should build locally.

**Comment by jachris** [comment]
I don't think it's a must-have. I'm personally fine with building locally or rebasing. But this would allow longer-running branches to use the CI artifacts without rewriting history due to rebasing. Also, the current behavior can be a little bit confusing if one is not aware of this issue.

**Comment by RalfJung** [comment]
> Also, the current behavior can be a little bit confusing if one is not aware of this issue.

Yeah it is confusing. :D Could you explain what happens?

**Comment by jachris** [comment]
From what I can tell, yes. The logic that determines that commit seems to not account for merge commits. But @jyn514, who helped me troublshoot this issue initially, might know the details.

**Comment by RalfJung** [comment]
I now also ran into this because syncing Miri changes back into rustc creates merge commits. Not sure how to work around this problem; rebasing is not an option for these syncs and a local build takes forever...

Cc @jyn514

**Comment by RalfJung** [comment]
Specifically that Miri history contains commits made by bors *in Miri*, which then fools the detection logic to think that was a rustc bors commit.

**Comment by RalfJung** [comment]
Currently, bootstrap uses the following command to determine the commit to fetch LLVM for
```
git rev-list --author=bors@rust-lang.org -n1 --first-parent HEAD -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
```
I can see two ways of making things work even for the Miri sync situation:
- Instead of `-n1`, use `-n10` or so and try several commits. For the case I have right now, the 2nd commit should already work.
- Instead of `HEAD`, use `$(git merge-base HEAD origin/master)`. However we can't really know what the `origin` remote is called, so it'd be more like
```
git fetch https://github.com/rust-lang/rust master
git merge-base HEAD FETCH_HEAD
```

**Comment by jachris** [comment]
I think the second option would be a solid solution, and it also works for the usecase I had in mind.

(Side note: Though I'm sure why this uses `--first-parent`. If we are on `master`, there should be no merge commits, and this has no effect. If we are on another branch, it discards merges into that branch... which is basically the original issue.)

**Comment by RalfJung** [comment]
On master there are tons of merge commits -- the ones created by bors.  In fact the commit we are looking for is always a merge commit, so maybe we should add `--merges`.

**Comment by jachris** [comment]
Ah yes, right, of course. The "no merge-commit policy" is only for PRs... Nevermind.

**Comment by Mark-Simulacrum** [comment]
Fetching introduces a network call on each x.py invocation, I think, since we need to know if something has changed. I guess we could try to cache based on the head sha, but I'm not generally a fan of that.

I think the right answer is probably to change bors to include a repository in the commits it generates more explicitly - perhaps the email could be bors+rust@rust-lang.org, for example.

Would need some care to update all of our automation, but seems achievable.

**Comment by RalfJung** [comment]
Yeah I agree having to invoke the network is not great. In principle we could search the user's remotes for one that points to our github repo and use `$remote/master`... but that seems pretty fragile.

Having different bors email addresses for different repos would help for the Miri case, yeah. And I guess for @jachris we could replace the `--first-parent` by `--merges`; the email filtering should be enough to trim down the candidates.

@Mark-Simulacrum I take it you are also not a fan of trying multiple commit candidates? I guess that would require caching negative lookups so it's also not great.

A temporary hack that would be helpful is some way to overwrite the `HEAD` commit from which it starts searching. For the Miri PR I just locally patched bootstrap to be able to do `x.py check` which I needed to update the lockfile.

**Comment by Mark-Simulacrum** [comment]
Updating the lockfile should be cargo metadata >/dev/null or something like that, no need for Cargo.

Yeah, multiple candidates I think has similar problems in terms of network hits (or negative cache of some kind).

--merges seems fine to me, but I'm not sure we can do away with --first-parent - for example, in the case of rollups, we need to be sure we're looking at the first parent commit to traverse the commit graph. I guess with -n1 that might not matter but it feels iffy to me.

Can you point me at the SHA/branch for the miri merges that are causing trouble? I wonder if the merge commits might be distinguishable by (for example) not having a tree associated with them that contains src/stage0.json or something like that.

**Comment by RalfJung** [comment]
Yeah you can just check out the branch for https://github.com/rust-lang/rust/pull/103392 to see the problem.

> I wonder if the merge commits might be distinguishable by (for example) not having a tree associated with them that contains src/stage0.json or something like that.

No that wouldn't help... this is Miri history reconstructed into the rust repo, it also has all of the rest of the rust repo around.

> Updating the lockfile should be cargo metadata >/dev/null or something like that, no need for x.py.

Ah, I never even tried doing anything outside of x.py in the rustc repo, it seemed hopeless. But anyway sometimes I might also have to actually build that branch.

**Comment by Mark-Simulacrum** [comment]
Yeah, I see. OK, I think the best option then is going to be the adjustment to bors to highlight which repository the merge commit is from. We'll want to check against triagebot, crater, rustc-perf, cargo-bisect-rustc, and the promote-release infrastructure which I think are the infrastructure places that have some amount of processing of rustc commit history.

**Comment by jyn514** [comment]
We don't need to fetch to support remotes named something other than `origin`, we can reuse the logic in https://github.com/rust-lang/rust/blob/3e565f1a27a19f7da48c7109500b4351c0819e68/src/tools/build_helper/src/git.rs#L22-L30

> * $(git merge-base HEAD origin/master)

👍 this seems like a good solution

**Comment by RalfJung** [comment]
Reopened by https://github.com/rust-lang/rust/pull/114345. I hope the perf team can help figure out what is going on; just calling `git merge-base` shouldn't cause any side-effects.

**Comment by RalfJung** [comment]
This regressed again. :(
```
downloading https://ci-artifacts.rust-lang.org/rustc-builds/23efae075aaf57714e8403572f151491869273fb/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404

ERROR: failed to download llvm from ci

    HELP: old builds get deleted after a certain time
    HELP: if trying to compile an old commit of rustc, disable `download-ci-llvm` in config.toml:

    [llvm]
    download-ci-llvm = false
```

**Comment by RalfJung** [comment]
It's picking up some bors commit from the Miri repo, I think.

**Comment by Swatinem** [comment]
I was also hit by this since quite a long time and was resorting to building LLVM from source.

Digging a bit deeper and `dbg!`-printing a couple of git commands, I stumbled across this comment here:

https://github.com/rust-lang/rust/blob/580fa0c1a9d4d5f90988c55bde7c77cdce52c3b1/src/tools/build_helper/src/git.rs#L82-L85

As indeed, the detection logic was using my local `upstream/master` branch which I haven’t touched since 2021 in fact, as I was instead using the GitHub UI to keep my own `origin/master` in sync.

I wonder if its possible to detect this somehow that the `rust-lang/rust` remote is hopelessly outdated?

**Comment by RalfJung** [comment]
Ah, that's annoying that Github guides users towards circumventing the normal Git workflows. :/

We could maybe detect when the detected `upstream/master` is more than a week old an run `git fetch`?

Though if bors was just putting a bit more information into the merge commit, none of this would be needed... these are all just hacks to work around not knowing which bors commits come from which repository.

**Comment by RalfJung** [comment]
Now that subrepos are not using bors any more, this should be much less of an issue -- the logic that searches for a "merge commit by bors" should no longer be confused by bors merge commits that were actually merging a Miri/clippy/... PR.

**Comment by RalfJung** [comment]
Hm, I'm still hitting this, trying to build a miri-sync branch -- it apparently tries to build LLVM, which then fails as the sources are not checked out. The bors commits here are all rust-lang/rust bors commits, but the history is a bit odd since the latest bors commits are in the *second* parent, not the first parent. The first parent is the Miri-side history.
```
f461981 2025-02-24 07:20 +0000 Ralf Jung                            M─┐ [miri-sync] {ralf/miri-sync} Merge pull request #4203 from RalfJung/rustup
2a8a9f8 2025-02-24 07:56 +0100 Ralf Jung                            │ M─┐ Merge from rustc
e0be1a0 2025-02-24 03:06 +0000 bors                                 │ │ M─┐ [master.2] [master] {origin/HEAD} {origin/master} Auto merge of #137271 - nikic:gep-nuw-2, r=scottmcm
9e7b184 2025-02-19 15:14 +0100 Nikita Popov                         │ │ │ o Also use gep inbounds nuw for index projections
31cc4c0 2025-02-19 11:06 +0100 Nikita Popov                         │ │ │ o Emit getelementptr inbounds nuw for pointer::add()
5e9d8a7 2025-02-19 10:55 +0100 Nikita Popov                         │ │ │ o Switch to the LLVMBuildGEPWithNoWrapFlags API
9af8985 2025-02-23 23:49 +0000 bors                                 │ │ M─│─┐ Auto merge of #137497 - tgross35:rollup-1oeclrr, r=tgross35
18ffee2 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─┐ Rollup merge of #137483 - bend-n:😅, r=Noratrieb
c813d8f 2025-02-23 20:26 +0700 bendn                                │ │ │ │ │ o rename sub_ptr 😅
e8342e7 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─┐ Rollup merge of #137452 - Shourya742:2025-02-23-add-module-level-doc-for-core-metadata, r=Kobzol
54dd4c8 2025-02-23 08:37 +0530 bit-aloo                             │ │ │ │ │ │ o bootstrap: add module docs for core:metadata
781203d 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─┐ Rollup merge of #137451 - compiler-errors:synm, r=Noratrieb
506532a 2025-02-23 03:02 +0000 Michael Goulet                       │ │ │ │ │ │ │ o The sym crate is not a thing
fb54acd 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─│─┐ Rollup merge of #137297 - tgross35:update-builtins, r=tgross35
08f1086 2025-02-19 20:17 -0500 Trevor Gross                         │ │ │ │ │ │ │ │ o Update `compiler-builtins` to 0.1.147
5f35f78 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─│─│─┐ Rollup merge of #137099 - yotamofek:pr/rustdoc/fix-ignored-test-directives, r=fmease
2c1f489 2025-02-15 21:12 +0000 Yotam Ofek                           │ │ │ │ │ │ │ │ │ o fix rustdoc test directives that were accidentally ignored replace "// @" with "//@ ", and fix the tests so they actually pass, after directives are checked
2c6fa32 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─│─│─│─┐ Rollup merge of #136637 - Pyr0de:binary-format, r=Noratrieb
a2bb4d7 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─│─│─│─│─┐ Rollup merge of #136543 - RalfJung:round-ties-even, r=tgross35
b9d0555 2025-02-23 10:57 +0100 Ralf Jung                            │ │ │ │ │ │ │ │ │ │ │ o add stdarch compatibility hack
d1b34ac 2025-02-22 14:12 +0100 Ralf Jung                            │ │ │ │ │ │ │ │ │ │ │ o make the new intrinsics safe
04e7a10 2025-02-04 15:15 +0100 Ralf Jung                            │ │ │ │ │ │ │ │ │ │ │ o intrinsics: unify rint, roundeven, nearbyint in a single round_ties_even intrinsic
31719b5 2025-02-23 14:30 -0500 Trevor Gross                         │ │ │ │ M─│─│─│─│─│─│─│─┐ Rollup merge of #136439 - yotamofek:pr/codegen-ssa-no-indexing, r=Noratrieb
73d5fe1 2025-02-02 13:09 +0000 Yotam Ofek                           │ │ │ │ │ │ │ │ │ │ │ │ o rustc_codegen_ssa: cleanup codegen attrs
a821785 2025-02-02 13:08 +0000 Yotam Ofek                           │ │ │ │ │ │ │ │ │ │ │ │ o rustc_codegen_ssa: simplify test for incompatible dependency formats thanks @kadiwa4
9e5e6a9 2025-02-02 13:04 +0000 Yotam Ofek                           │ │ │ │ │ │ │ │ │ │ │ │ o rustc_codegen_ssa: cleanup nested `if`s and a needless `match`
c930bb2 2025-02-02 13:01 +0000 Yotam Ofek                           │ │ │ │ │ │ │ │ │ │ │ │ o rustc_codegen_ssa: use slice patterns instead of len-check+indexing
f8a913b 2025-02-23 20:32 +0000 bors                                 │ │ M─│─│─│─│─│─│─│─│─│─│─┐ Auto merge of #137476 - onur-ozkan:137469, r=jieyouxu
```

Cc @onur-ozkan any idea what we could do about this?

**Comment by onur-ozkan** [comment]
I assume if you configure upstream master, it should work fine then.

**Comment by RalfJung** [comment]
What do you mean by that? I am working on the `miri-sync` branch so of course the "uptstream" branch for git is miri-sync. That's a standard feature branch setup. Of course I do have a remote called `origin` that has an up-to-date upstream master in it, but it is not reasonable to expect my current branch to have rust-lang/rust master as its upstream.

**Comment by onur-ozkan** [comment]
What I mean adding a remote that is called "upstream". `git remote add upstream https://github.com/rust-lang/rust` should do it.

We do the same on CI at https://github.com/rust-lang/rust/blob/e0be1a02626abef2878cb7f4aaef7ae409477112/src/ci/scripts/setup-upstream-remote.sh#L18

for jobs that are running outside of rust-lang/rust repository.

**Comment by RalfJung** [comment]
Uh, why should I call it "upstream"? The standard name is "origin" isn't it?

**Comment by RalfJung** [comment]
Anyway I think the main question is whether the `--first-parent` in that logic is load-bearing or whether we can remove it.

**Comment by RalfJung** [comment]
> I think that situation is relatively rare, but would e.g. come up with subtree syncs.

A Miri subtree sync is exactly where `--first-parent` is causing problems. :)

**Comment by Mark-Simulacrum** [comment]
@RalfJung Can you say more about how you ended up with f4619816b8cc9e37a5b2f5a5a6a89ce5117c4878 as HEAD? By the first parent history that traces to a bors commit on Tue Oct 22 08:21:37 2024 (quite old...), which makes sense in terms of why that broke things.

For bors merges, we take care that the first parent of the merge commit is the previous bors merge. This property is also kept for rollup merges (e.g., 0a33d7ccd1491b9c751624bae9802e2578e3c067 has a first parent of 353650332d8 which ends up chaining recursively to c49fc911f451b6e27dcd7fbfac77ab2902fd3ae5 and finally a9e7b30487235621751cc628f170c0f15fb215c4 for that rollup's bors merge parent).

It seems like miri is *not* respecting that property, sometimes the "rustc side" of the merge is most recent through the 2nd parent commit -- which is why `--first-parent` is causing issues here. If that property was respected, then first-parent would work fine.

The reason for first-parent (or something like it -- doesn't have to be first necessarily) is that the alternative (most recent commit ("reverse chronological order" by docs in rev-list) with bors@rust-lang.org as the email) *could have* found a clippy, miri, etc. bors merge commit as being more recent in the past, even in rust-lang/rust history, if you were syncing in merge commits from those repos (first-parent avoided that there because it refused to traverse the "sideways" history). But now that we no longer use bors anywhere except rustc, it seems plausible that just dropping `--first-parent` is actually fine? The most recent bors@rust-lang.org commit you can find *should* always be the right one to use, since nothing else is producing those commits other than rust-lang/rust CI these days IIUC...

It seems like you came to that conclusion earlier (https://github.com/rust-lang/rust/issues/101907#issuecomment-2466170296) but we didn't actually follow up and improve the finding logic given that realization :)

**Comment by RalfJung** [comment]
> Can you say more about how you ended up with https://github.com/rust-lang/rust/commit/f4619816b8cc9e37a5b2f5a5a6a89ce5117c4878 as HEAD? By the first parent history that traces to a bors commit on Tue Oct 22 08:21:37 2024 (quite old...), which makes sense in terms of why that broke things.

It was created by `./miri rustc-push` in the Miri repo. IOW, this is a standard Miri subtree sync. The way those come out, the PR itself contains a merge commit (which we usually forbid but [have to] allow for subtree syncs) where the first parent is the Miri history and the 2nd parent is the rustc history. By going up the first parents, you find an old bors commit created in the Miri repo, from the time when Miri used bors.

**Comment by RalfJung** [comment]
Somehow the removal of `--first-parent` caused https://github.com/rust-lang/rust/pull/137661 to fail so it had to be reverted. Does anyone understand why? That PR doesn't even contain a merge commit. It also changes LLVM so it shouldn't download anything anyway, right?

Not being able to test Miri syncs locally is quite annoying so it'd be really good to finally resolve this issue.

**Comment by onur-ozkan** [comment]
> Somehow the removal of --first-parent caused https://github.com/rust-lang/rust/pull/137661 to fail so it had to be reverted. Does anyone understand why? That PR doesn't even contain a merge commit. It also changes LLVM so it shouldn't download anything anyway, right?

Yeah, it shouldn't. I am planning to have a look at that (both on git and LLVM download logics) in a couple of days.

**Comment by Mark-Simulacrum** [comment]
> That PR doesn't even contain a merge commit

Note that CI *does* run on a merge commit (f43d771d2eafb5ae9b94f2734073c91e213320db in the case of the failed CI run).

However both parents of the merge commit are recent (a46c755c6ad, the first parent, looks right to me), it is also the indirect parent of the 2nd parent (e045a025661358ca83887b68a7069b9143202ca1 -> a46c755c6ad).

An *empty* rev-list result so far for me is from this[^1][^2]:

```
$ git rev-list --first-parent --author=bors@rust-lang.org -n1 f43d771d2eafb5ae9b94f2734073c91e213320db -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
a46c755c6ad2c9543a836a1ca134178b2345acc2

$ git rev-list --author=bors@rust-lang.org -n1 f43d771d2eafb5ae9b94f2734073c91e213320db -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
# empty
```

[^1]: Note that I thought originally CI was using `git merge-base origin/master HEAD` to compute the commit to pass to rev-list (see code: https://github.com/rust-lang/rust/blob/0c72c0d11adeba449886089c6bd5d48363f7a2cd/src/build_helper/src/git.rs#L133-153), but this does not currently run. src/ci/channel is "nightly\n", not "nightly", and we are in a rust-lang managed CI job.
[^2]: Also note that neither result ends up leading to a correct outcome. This PR CI should **not** download LLVM, because it is modifying LLVM. We refuse to do that if the [detected LLVM SHA is the same as HEAD](https://github.com/rust-lang/rust/blob/8c392966a013fd8a09e6b78b3c8d6e442bc278e1/src/bootstrap/src/core/build_steps/llvm.rs#L285-L290), but in the case of **pull request** CI, the HEAD sha is not a bors merge, so we'd happily download some other LLVM AFAICT, not actually testing the changes until the change hits proper bors CI.

It's possible my local env is messed up, but if I set it up correctly, all the "known" commits are:

```
$ git rev-list f43d771d2eafb5ae9b94f2734073c91e213320db
f43d771d2eafb5ae9b94f2734073c91e213320db
e045a025661358ca83887b68a7069b9143202ca1
a46c755c6ad2c9543a836a1ca134178b2345acc2
```

(because we have a shallow fetch: `2025-02-26T08:52:41.3655527Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=2 origin +f43d771d2eafb5ae9b94f2734073c91e213320db:refs/remotes/pull/137661/merge`)

That said, it's not very clear to me how first-parent managed to find a commit when regular traversal doesn't. I wouldn't expect that to increase what results are available to git. Maybe the diff attributed to a commit changes depending on how that commit is reached, given the shallow history?

**Comment by RalfJung** [comment]
Uh, removing --first-parent finds fewer things? Yeah that makes no sense to me.

Can we just treat "empty result" as "do not download LLVM"?

**Comment by RalfJung** [comment]
> Note that CI does run on a merge commit (https://github.com/rust-lang/rust/commit/f43d771d2eafb5ae9b94f2734073c91e213320db in the case of the failed CI run).

Where does that commit come from? It says @nikic created it, but it doesn't show up in the PR itself...?

Is this the commit created by github as part of CI? I didn't realize that would be attributed to the PR author, interesting.

> An empty rev-list result so far for me is from this

How do you reproduce this? Did you do a shallow clone and then get that result?

Locally when I run that comment on my regular clone, I get
```
$ git rev-list --author=bors@rust-lang.org -n1 f43d771d2eafb5ae9b94f2734073c91e213320db -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
ce36a966c79e109dabeef7a47fe68e5294c6d71e
```
This is the same result I also get with `--first-parent`. However, that commit is quite a ways back in the history; a `--depth=2` checkout would not contain it. But that would always happen, right? A `--depth=2` clone will not contain the last commit that changed LLVM most of the time.

The main thing I am wondering about how: given that we have a shallow clone in CI, it makes no sense at all to try and determine anything about the git history. What even is our plan here, what is the intended behavior?

Also, what is the intended behavior if LLVM has changed since the most recent merge commit that updates LLVM? It seems the current logic would just blatantly ignore that and use the downloaded LLVM anyway, or is there a check somewhere that I missed?

**Comment by RalfJung** [comment]
> How do you reproduce this? Did you do a shallow clone and then get that result?

Okay yeah I can reproduce this as well, using
```
git init rustc
cd rustc
git remote add origin https://github.com/rust-lang/rust
git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=2 origin +f43d771d2eafb5ae9b94f2734073c91e213320db:refs/remotes/pull/137661/merge
```

The answer you get with `--first-parent` is still wrong: it's just the current master commit (i.e. what got merged into the PR commit to form the CI commit), no matter whether it changed LLVM or not.

Comparing `git rev-list f43d771d2eafb5ae9b94f2734073c91e213320db -- src/llvm-project` and `git rev-list --first-parent f43d771d2eafb5ae9b94f2734073c91e213320db -- src/llvm-project`, they produce disjoint sets of commits. The docs for `--first-parent` mention something interesting:

> This option also changes default diff format for merge commits to first-parent, see --diff-merges=first-parent for details.

And indeed, using `--diff-merges=first-parent` instead of `--first-parent` does produce a non-empty result:
```
git rev-list --author=bors@rust-lang.org -n1 f43d771d2eafb5ae9b94f2734073c91e213320db --diff-merges=first-parent -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
a46c755c6ad2c9543a836a1ca134178b2345acc2
```
And it seems like that option will also work for Miri sync commits:
```
$ git rev-list --author=bors@rust-lang.org -n1 716dd22 --diff-merges=first-parent -- src/llvm-project src/bootstrap/download-ci-llvm-stamp src/version
07697360aee0cebcb4e304236ba1884d8dde5469
```

However, I am still not convinced it makes sense to run `git_upstream_merge_base` at all in a shallow clone, as the commit you're getting out will always just be the most recent bors commit, no matter whether it changed LLVM or not.


> This would silently hide actual bugs in the commit finding logic.

We are already silently hiding bugs in that logic, it seems. I'd rather silently do the right thing, then. ;) There is nothing useful that can be gleamed

**Comment by RalfJung** [comment]
FWIW the points about shallow clones could also explain this:
```
            // FIXME: When running on rust-lang managed CI and it's not a nightly build,
            // `git_upstream_merge_base` fails with an error message similar to this:
```
with such a shallow view of what the history is, there's no way git can reliably compute the merge base.

On nightly CI, `git merge-base` still works since the very latest master commit *is* part of the shallow clone (it is the first parent of HEAD). On beta/stable CI, ofc the merge base is further away, and therefore not part of the shallow clone any more, so `git merge-base` fails.

So we have to either make a full clone, or stop doing any kind of git history queries in CI.

**Comment by Skgland** [comment]
> So we have to either make a full clone, or stop doing any kind of git history queries in CI. (But also, on beta/stable CI, I don't think the merge base with master should even be relevant.)

Would a [partial (blobless) clone](https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone) instead of a shallow clone work maybe?

**Comment by RalfJung** [comment]
I don't think that would help since we are filtering based on the diff of the commits, and determining the diff needs the blobs.

**Comment by Skgland** [comment]
> I don't think that would help since we are filtering based on the diff of the commits, and determining the diff needs the blobs.

Right, I was for some reason thinking that filtering was based on commit message content.
In that case blobless clone should cause it to download blobs one by one for all commits it tries to diff.
While it might fix the problem, loading blobs one by one sounds undesirable.

**Comment by jieyouxu** [comment]
Reopening because I'm posting a revert for #137593 to fix perf.

**Comment by RalfJung** [comment]
Where can I see the failing CI log that is fixed by the revert?

**Comment by jieyouxu** [comment]
(Asking in [#t-infra > benchmarks failed?](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/benchmarks.20failed.3F/with/507663506))
